File: android/src/main/kotlin/com/eopeter/fluttermapboxnavigation/models/MarkerConfiguration.kt
package com.eopeter.fluttermapboxnavigation.models

/**
 * Configuration for static marker display and behavior
 */
data class MarkerConfiguration(
    val showDuringNavigation: Boolean = true,
    val showInFreeDrive: Boolean = true,
    val showOnEmbeddedMap: Boolean = true,
    val maxDistanceFromRoute: Double? = null,
    val minZoomLevel: Double = 10.0,
    val enableClustering: Boolean = true,
    val maxMarkersToShow: Int? = null,
    val defaultIconId: String? = null,
    val defaultColor: Int? = null
) {
    companion object {
        /**
         * Creates a MarkerConfiguration from a JSON map
         */
        fun fromJson(json: Map<String, Any>?): MarkerConfiguration {
            if (json == null) return MarkerConfiguration()
            
            return MarkerConfiguration(
                showDuringNavigation = json["showDuringNavigation"] as? Boolean ?: true,
                showInFreeDrive = json["showInFreeDrive"] as? Boolean ?: true,
                showOnEmbeddedMap = json["showOnEmbeddedMap"] as? Boolean ?: true,
                maxDistanceFromRoute = json["maxDistanceFromRoute"] as? Double,
                minZoomLevel = (json["minZoomLevel"] as? Number)?.toDouble() ?: 10.0,
                enableClustering = json["enableClustering"] as? Boolean ?: true,
                maxMarkersToShow = json["maxMarkersToShow"] as? Int,
                defaultIconId = json["defaultIconId"] as? String,
                defaultColor = json["defaultColor"] as? Int
            )
        }
    }

    /**
     * Converts the MarkerConfiguration to a JSON map
     */
    fun toJson(): Map<String, Any> {
        val json = mutableMapOf<String, Any>(
            "showDuringNavigation" to showDuringNavigation,
            "showInFreeDrive" to showInFreeDrive,
            "showOnEmbeddedMap" to showOnEmbeddedMap,
            "minZoomLevel" to minZoomLevel,
            "enableClustering" to enableClustering
        )

        maxDistanceFromRoute?.let { json["maxDistanceFromRoute"] = it }
        maxMarkersToShow?.let { json["maxMarkersToShow"] = it }
        defaultIconId?.let { json["defaultIconId"] = it }
        defaultColor?.let { json["defaultColor"] = it }

        return json
    }

    override fun toString(): String {
        return "MarkerConfiguration(" +
                "showDuringNavigation=$showDuringNavigation, " +
                "showInFreeDrive=$showInFreeDrive, " +
                "showOnEmbeddedMap=$showOnEmbeddedMap, " +
                "maxDistanceFromRoute=$maxDistanceFromRoute, " +
                "minZoomLevel=$minZoomLevel, " +
                "enableClustering=$enableClustering, " +
                "maxMarkersToShow=$maxMarkersToShow)"
    }
} 

File: lib/src/models/marker_configuration.dart
import 'package:flutter/material.dart';
import 'static_marker.dart';

/// Configuration for static marker display and behavior
class MarkerConfiguration {
  /// Whether to show markers during navigation
  final bool showDuringNavigation;
  
  /// Whether to show markers in free drive mode
  final bool showInFreeDrive;
  
  /// Whether to show markers on embedded map view
  final bool showOnEmbeddedMap;
  
  /// Maximum distance in kilometers from the current route to show markers
  /// If null, all markers will be shown regardless of route distance
  final double? maxDistanceFromRoute;
  
  /// Minimum zoom level to show markers (markers hidden when zoomed out too far)
  final double minZoomLevel;
  
  /// Whether to enable marker clustering for dense areas
  final bool enableClustering;
  
  /// Maximum number of markers to display at once (for performance)
  final int? maxMarkersToShow;
  
  /// Callback function when a marker is tapped
  final Function(StaticMarker)? onMarkerTap;
  
  /// Default icon ID to use when marker doesn't specify one
  final String? defaultIconId;
  
  /// Default color to use when marker doesn't specify one
  final Color? defaultColor;

  const MarkerConfiguration({
    this.showDuringNavigation = true,
    this.showInFreeDrive = true,
    this.showOnEmbeddedMap = true,
    this.maxDistanceFromRoute,
    this.minZoomLevel = 10.0,
    this.enableClustering = true,
    this.maxMarkersToShow,
    this.onMarkerTap,
    this.defaultIconId,
    this.defaultColor,
  });

  /// Creates a MarkerConfiguration from a JSON map
  factory MarkerConfiguration.fromJson(Map<String, dynamic> json) {
    return MarkerConfiguration(
      showDuringNavigation: json['showDuringNavigation'] as bool? ?? true,
      showInFreeDrive: json['showInFreeDrive'] as bool? ?? true,
      showOnEmbeddedMap: json['showOnEmbeddedMap'] as bool? ?? true,
      maxDistanceFromRoute: json['maxDistanceFromRoute'] as double?,
      minZoomLevel: json['minZoomLevel'] as double? ?? 10.0,
      enableClustering: json['enableClustering'] as bool? ?? true,
      maxMarkersToShow: json['maxMarkersToShow'] as int?,
      defaultIconId: json['defaultIconId'] as String?,
      defaultColor: json['defaultColor'] != null 
          ? Color(json['defaultColor'] as int) 
          : null,
    );
  }

  /// Converts the MarkerConfiguration to a JSON map
  Map<String, dynamic> toJson() {
    return {
      'showDuringNavigation': showDuringNavigation,
      'showInFreeDrive': showInFreeDrive,
      'showOnEmbeddedMap': showOnEmbeddedMap,
      'maxDistanceFromRoute': maxDistanceFromRoute,
      'minZoomLevel': minZoomLevel,
      'enableClustering': enableClustering,
      'maxMarkersToShow': maxMarkersToShow,
      'defaultIconId': defaultIconId,
      'defaultColor': defaultColor?.value,
    };
  }

  /// Creates a copy of this configuration with updated fields
  MarkerConfiguration copyWith({
    bool? showDuringNavigation,
    bool? showInFreeDrive,
    bool? showOnEmbeddedMap,
    double? maxDistanceFromRoute,
    double? minZoomLevel,
    bool? enableClustering,
    int? maxMarkersToShow,
    Function(StaticMarker)? onMarkerTap,
    String? defaultIconId,
    Color? defaultColor,
  }) {
    return MarkerConfiguration(
      showDuringNavigation: showDuringNavigation ?? this.showDuringNavigation,
      showInFreeDrive: showInFreeDrive ?? this.showInFreeDrive,
      showOnEmbeddedMap: showOnEmbeddedMap ?? this.showOnEmbeddedMap,
      maxDistanceFromRoute: maxDistanceFromRoute ?? this.maxDistanceFromRoute,
      minZoomLevel: minZoomLevel ?? this.minZoomLevel,
      enableClustering: enableClustering ?? this.enableClustering,
      maxMarkersToShow: maxMarkersToShow ?? this.maxMarkersToShow,
      onMarkerTap: onMarkerTap ?? this.onMarkerTap,
      defaultIconId: defaultIconId ?? this.defaultIconId,
      defaultColor: defaultColor ?? this.defaultColor,
    );
  }

  @override
  String toString() {
    return 'MarkerConfiguration('
        'showDuringNavigation: $showDuringNavigation, '
        'showInFreeDrive: $showInFreeDrive, '
        'showOnEmbeddedMap: $showOnEmbeddedMap, '
        'maxDistanceFromRoute: $maxDistanceFromRoute, '
        'minZoomLevel: $minZoomLevel, '
        'enableClustering: $enableClustering, '
        'maxMarkersToShow: $maxMarkersToShow)';
  }
} 